### PR TITLE
Refactor plot autosize

### PR DESCRIPTION
**Pull request by etpinard**
fixes https://github.com/plotly/plotly.js/issues/537

This PR sets up a base branch for the in-progress autosize refactor starting out by @n-riesco in #577 patched in #629 and then reverted in #633.

Important comments:

- https://github.com/plotly/plotly.js/pull/629#issuecomment-225591388
- https://github.com/plotly/plotly.js/pull/577#discussion_r66688752

### CODE REVIEW

File: src/lib/index.js [hunk 1]
```diff
@@ -379,7 +379,9 @@ lib.getPlotDiv = function(el) {
 
 lib.isPlotDiv = function(el) {
     var el3 = d3.select(el);
-    return el3.size() && el3.classed('js-plotly-plot');
+    return el3.node() instanceof HTMLElement &&
```
Comment (mdtusz) [inline src/lib/index.js]: Just for my own understanding: when would this _not_ be an HTMLElement?

File: src/plot_api/plot_api.js [hunk 2]
```diff
@@ -2172,14 +2166,8 @@ Plotly.relayout = function relayout(gd, astr, val) {
...
-        // check autosize or autorange vs size and range
-        if(hw.indexOf(ai) !== -1) {
-            doextra('autosize', false);
-        }
-        else if(ai === 'autosize') {
-            doextra(hw, undefined);
-        }
-        else if(pleafPlus.match(/^[xyz]axis[0-9]*\.range(\[[0|1]\])?$/)) {
+        // check autorange vs range
+        if(pleafPlus.match(/^[xyz]axis[0-9]*\.range(\[[0|1]\])?$/)) {
```
Comment (mdtusz) [inline src/plot_api/plot_api.js]: Should this not exclude `xaxis0` and `xaxis1`? (And similarly `y` and `z`...)

File: src/plots/layout_attributes.js [hunk 3]
```diff
@@ -45,13 +45,17 @@ module.exports = {
...
-        // TODO: better handling of 'initial'
-        values: [true, false, 'initial'],
+        dflt: false,
         description: [
-            'Determines whether or not the dimensions of the figure are',
-            'computed as a function of the display size.'
+            'Determines whether or not a layout width or height',
+            'that has been left undefined by the user',
+            'is initialized on each relayout.',
+
```
Comment (mdtusz) [inline src/plots/layout_attributes.js]: :cow2:?

File: src/lib/index.js (hunk 1 above)
Comment (etpinard) [inline src/lib/index.js]: if someone calls `Plotly.Plots.supplyDefaults({ data: [], layout: {}});` which should be allowed.

File: src/plots/plots.js [hunk 4]
```diff
@@ -752,6 +787,93 @@ plots.supplyLayoutGlobalDefaults = function(layoutIn, layoutOut) {
...
 
+plots.plotAutoSize = function plotAutoSize(gd, layout, fullLayout) {
+    var context = gd._context || {},
+        frameMargins = context.frameMargins,
+        newWidth,
+        newHeight;
+
+    var isPlotDiv = Lib.isPlotDiv(gd);
+
+    if(isPlotDiv) gd.emit('plotly_autosize');
```
Comment (mdtusz) [inline src/plots/plots.js]: Should we emit this here, or after the resizing?

File: src/lib/index.js (hunk 1 above)
Comment (n-riesco) [inline src/lib/index.js]: If `gd` isn't a `HTMLElement`, `el3.classed('js-plotly-plot')` throws an exception.

File: src/plots/plots.js (hunk 4 above)
Comment (etpinard) [inline src/plots/plots.js]: It has to be here, at least for now, for backward compat in the old plot.ly workspace.

File: src/plots/layout_attributes.js (hunk 3 above)
Comment (n-riesco) [inline src/plots/layout_attributes.js]: Yes, this wording could be improved. The most important points I wanted to make in the description were that:
- if the user supplies values of `layout.width` and/or `layout.height`, `plotAutoSize` doesn't change them.
- if `layout.autosize` is false, `plotAutosize` is called once.
- if `layout.autosize` is true, `plotAutosize` is called on each relayout.

Comment (mdtusz) [inline src/plots/layout_attributes.js]: I was meaning just the blank line. I think the description itself is clear enough.

Comment (n-riesco) [inline src/plots/layout_attributes.js]: T'was intentional :smile:  but if it looks like a :cow:, it must be one.

**Comment by mdtusz** [comment]
💃

**Comment by etpinard** [comment]
dropping this PR from the `v1.17.0` milestone because of the numerous merge conflict left to fix ... and let's face it, we're due for a minor bump.

**Comment by fresheneesz** [comment]
@etpinard 

I created some logic that finds the maximum size an element can be sized to without exceeding the boundaries of its parent. It takes into account margins, borders, padding, percentage widths, max-height/width, and sibling nodes. You can find it here: http://stackoverflow.com/a/39782139/122422

I think this is something that would help autosizing work better, since currently using autosizing seems to frequently exceed the boundaries of the graph's parent.

**Comment by etpinard** [comment]
Dropping from the `v1.18.0` milestone.

This will be implemented in `v1.19.0` together with layout breakpoints.

**Comment by JonDum** [comment]
I'm super interested in this PR. Is there anything I can contribute? 

I am trying to use Plotly in a Flex layout and Plotly is not playing nicely. The major problem is Plotly's use of explicit width/height in pixels. Doing this makes an implicit size on the chart that will then cause the flex layout to never go less than size. 

Here's an example pen: http://codepen.io/JonDum/pen/gwEPPA

If you resize horizontally you'll notice that it will grow larger, but never smaller.

If you look at how [Chartist.js](https://github.com/gionkunz/chartist-js/blob/develop/src/scripts/core.js#L300) handles this, they default to setting the `<svg>` element to a width/height of 100% and they automatically scale to their container's size (which you can see on their [demo page](https://gionkunz.github.io/chartist-js/))

**Comment by etpinard** [comment]
@n-riesco I (finally) resurrected your PR :tada: 

To preserve backward compatibility, I had to add two commits:

- https://github.com/plotly/plotly.js/pull/635/commits/77971f4d74d4e2654162f1c6a1d06494aeb6b1c0 add a hacky block in relayout to make sure [this](https://github.com/plotly/plotly.js/blob/ad864c461c1dc9f10b4a1214b3f74d5fa845229e/test/jasmine/tests/plot_api_test.js#L118-L133) relayout case still passes. It's not pretty, but only a few lines. I hope you don't mind.

- https://github.com/plotly/plotly.js/pull/635/commits/b6c56103f2ee64cb6eb27b7565cb0a100751e9f8 is a little trickier and very important. This commit ensure that graphs with no set `width` and `height` have a default autosize of `true`. Previously graphs with no set `width` and `height` were given the `'initial'` autosize value (see [here](https://github.com/plotly/plotly.js/blob/ad864c461c1dc9f10b4a1214b3f74d5fa845229e/src/plots/plots.js#L869-L870)).

I hope you find these two commits compatible with what you had in mind for this PR.

**Comment by n-riesco** [comment]
Looks like github didn't get my email. Here's a copy:

---

On 26/10/16 18:50,  Étienne Tétreault-Pinard wrote:
>   * 77971f4 

I don't understand how this commit works.

If I read it correctly, something like 'Plotly.relayout(gd, { width: null })` would be equivalent to 'Plotly.relayout(gd, { width: 'initial' })`.

Since 'initial' is an invalid width, this would make `gd._fullLayout.width = 700` (i.e. the default width value).

Would the following work instead?

```
if(['width', 'height'].indexOf(ai) !== -1 && vi === null) {
    delete gd._fullLayout._initialAutoSizeIsDone;
}
```

I can look into this tomorrow.

---

>   * b6c5610 

I think this behaviour is better (and necessary to fix #537)

**Comment by etpinard** [comment]
> Would the following work instead?
> 
> if(['width', 'height'].indexOf(ai) !== -1 && vi === null) {
>     delete gd._fullLayout._initialAutoSizeIsDone;
> }

Let me try. Thanks! 




> I think this behaviour is better

By _this_, you meant behaviour after my last commit (b6c5610) ?

> (and necessary to fix #537)

The tests you added to fix #537 are still passing, anything else I care to confirm that #537 is still fixed?

**Comment by etpinard** [comment]
@n-riesco 

> > Would the following work instead?

> > if(['width', 'height'].indexOf(ai) !== -1 && vi === null) {
delete gd._fullLayout._initialAutoSizeIsDone;
}

>Let me try. Thanks!

Replacing https://github.com/plotly/plotly.js/commit/77971f4d74d4e2654162f1c6a1d06494aeb6b1c0 with your suggestion makes the aforementioned relayout test [case](https://github.com/plotly/plotly.js/blob/ad864c461c1dc9f10b4a1214b3f74d5fa845229e/test/jasmine/tests/plot_api_test.js#L118-L133) fail.

The reason being:

- clearing `_fullLayout._initialAutoSizeDone` leads to a `plotAutosize` call which (in the default context) reaches this [line](https://github.com/plotly/plotly.js/pull/635/files#diff-ad4f76ccd6044ed16514297078e13b84R982) where the _new width/height_ are taken from the DOM
- The problem is: those width/height value come from the previous relayout call, so line by line:

```js
        it('sets null values to their default', function(done) {
            var defaultWidth;
            Plotly.plot(gd, [{ x: [1, 2, 3], y: [1, 2, 3] }])
                .then(function() {
                    defaultWidth = gd._fullLayout.width;
                    defaultWidth // => 700
                    return Plotly.relayout(gd, { width: defaultWidth - 25});
                })
                .then(function() {
                    gd._fullLayout.width // => 675
                    expect(gd._fullLayout.width).toBe(defaultWidth - 25);
                    return Plotly.relayout(gd, { width: null });
                })
                .then(function() {
                    gd._fullLayout.width // => 675 BUT should 700 !!
                    expect(gd._fullLayout.width).toBe(defaultWidth);
                })
                .then(done);
        });
```

**Comment by etpinard** [comment]
Now, by setting `width: 'initial'` in relayout in the hacky block of https://github.com/plotly/plotly.js/commit/77971f4d74d4e2654162f1c6a1d06494aeb6b1c0, `Plots.supplyDefaults` now reaches this [line](https://github.com/plotly/plotly.js/pull/635/files#diff-ad4f76ccd6044ed16514297078e13b84R389), where `newLayout.width` isn't reset to its old value as `'initial'` is truthy. Instead, `newLayout.width` is set back to its default value of 700.

**Comment by etpinard** [comment]
Perhaps a better solution would be to cache the initial graph div width / height computed styles and use them in `relayout`.

**Comment by n-riesco** [comment]
Another idea (but needs tweaking):

```js
if(['width', 'height'].indexOf(ai) !== -1 && vi === null) {
    delete gd._fullLayout._initialAutoSizeIsDone;
    delete.layout[ai];
    coerce(ai);
}
```

**Comment by etpinard** [comment]
@n-riesco here's a less hacky (I think) fix: https://github.com/plotly/plotly.js/commit/839601891b5251c613f671b192cc0484044e6128

Let me know what you think.

**Comment by etpinard** [comment]
@n-riesco commit https://github.com/plotly/plotly.js/commit/08069a4846808571066ce3d99b30bdb93b76161c implements an even less hacky solution (I think we now have the **correct** answer to what `relayout` should do).

Let me know what you think.

**Comment by etpinard** [comment]
@n-riesco I took your :+1: reaction on https://github.com/plotly/plotly.js/pull/635#issuecomment-256502896 as a sign of approval. Commit https://github.com/plotly/plotly.js/pull/635/commits/5fb7cd185afbc938f7ceb0f24ed551ae16118316 implement the cache-initial-autosize solution.

**Comment by n-riesco** [comment]
On 26/10/16 18:50,  Étienne Tétreault-Pinard wrote:
>   * 77971f4 <https://github.com/plotly/plotly.js/pull/635/commits/77971f4d74d4e2654162f1c6a1d06494aeb6b1c0> add a hacky block in relayout to make sure this <https://github.com/plotly/plotly.js/blob/ad864c461c1dc9f10b4a1214b3f74d5fa845229e/test/jasmine/tests/plot_api_test.js#L118-L133> relayout case still passes. It's not pretty, but only a few lines. I hope you don't mind.


I don't understand how this commit works.

If I read it correctly, something like 'Plotly.relayout(gd, { width: null })` would be equivalent to 'Plotly.relayout(gd, { width: 'initial' })`.

Since 'initial' is an invalid width, this would make `gd._fullLayout.width = 700` (i.e. the default width value).

Would the following work instead?

```
if(['width', 'height'].indexOf(ai) !== -1 && vi === null) {
    delete gd._fullLayout._initialAutoSizeIsDone;
}
```

I can look into this tomorrow.



>   * b6c5610 <https://github.com/plotly/plotly.js/pull/635/commits/b6c56103f2ee64cb6eb27b7565cb0a100751e9f8> is a little trickier and very important. This commit ensure that graphs with no set |width| and |height| have a default autosize of |true|. Previously graphs with no set |width| and |height| were given the |'initial'| autosize value (see here <https://github.com/plotly/plotly.js/blob/ad864c461c1dc9f10b4a1214b3f74d5fa845229e/src/plots/plots.js#L869-L870>).


I think this behaviour is better (and necessary to fix #537)

**Pull request merged**